### PR TITLE
Feature/add quality profile

### PR DESCRIPTION
**Pull request by kdwils**
Still needs tests written, but wanted overall feedback on the set up. If you delete your local db and rerun the server it'll populate with some defaults now.

You can test grabbing profiles with 2 new endpoints:

`http://localhost:8080/api/v1/quality/profiles` to dump back all profiles
`http://localhost:8080/api/v1/quality/profiles/1` to dump back profile with id 1

The high level quality_profile table basically defines a name and an ID for the profile
The quality_profile_item table maps a profile to the items in the profile
The quality_item table maps an item to a quality definition
The quality definition has the sauce

The parentID on the items was supposed to be if they were children of another item, but I wasn't sure if were going to need something like that. AI generated the default schema values for me for testing so they're probably not all that accurate..

The schema is pretty similar to other schemas we looked at, but is more table oriented instead of being a json field in the db.

Example Responses:
`http://localhost:8080/api/v1/quality/profiles/2`

```json
{
  "response": {
    "name": "High Definition",
    "items": [
      {
        "parentID": null,
        "name": "Bluray-1080p",
        "quality": {
          "name": "Bluray-1080p",
          "type": "movie",
          "preferredSize": 1999,
          "minSize": 50.8,
          "maxSize": 2000
        },
        "id": 8,
        "allowed": true
      },
      {
        "parentID": null,
        "name": "Bluray-720p",
        "quality": {
          "name": "HDTV-1080p",
          "type": "movie",
          "preferredSize": 1999,
          "minSize": 33.8,
          "maxSize": 2000
        },
        "id": 5,
        "allowed": true
      },
      {
        "parentID": null,
        "name": "WEBDL-720p",
        "quality": {
          "name": "Bluray-720p",
          "type": "movie",
          "preferredSize": 1999,
          "minSize": 25.7,
          "maxSize": 2000
        },
        "id": 4,
        "allowed": true
      },
      {
        "parentID": null,
        "name": "HDTV-720p",
        "quality": {
          "name": "WEBRip-720p",
          "type": "movie",
          "preferredSize": 1999,
          "minSize": 12.5,
          "maxSize": 2000
        },
        "id": 3,
        "allowed": true
      },
      {
        "parentID": null,
        "name": "HDTV-1080p",
        "quality": {
          "name": "WEBDL-1080p",
          "type": "movie",
          "preferredSize": 1999,
          "minSize": 12.5,
          "maxSize": 2000
        },
        "id": 6,
        "allowed": true
      },
      {
        "parentID": null,
        "name": "WEBDL-1080p",
        "quality": {
          "name": "WEBRip-1080p",
          "type": "movie",
          "preferredSize": 1999,
          "minSize": 12.5,
          "maxSize": 2000
        },
        "id": 7,
        "allowed": true
      }
    ],
    "id": 2,
    "cutoff": 8,
    "upgradeAllowed": true
  }
}
```

### CODE REVIEW

File: pkg/storage/sqlite/schema/schema.sql [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    "preferred_size" NUMERIC NOT NULL,
+    "min_size" NUMERIC NOT NULL,
+    "max_size" NUMERIC NOT NULL,
+    "media_type" TEXT NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS "quality_profile" (
+    "id" INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+    "name" TEXT NOT NULL,
+    "cutoff" INTEGER NOT NULL,
```
Comment (kasuboski) [inline pkg/storage/sqlite/schema/schema.sql:22]: This cutoff is referencing the quality_id of a quality_definition? Can we name it like `cutoff_quality_id` or something of that nature

File: pkg/storage/sqlite/schema/defaults.sql [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+(13, 'Bluray-2160p', 1999, 102, 2000, 'movie'),
+(14, 'Remux-2160p', 1999, 187.4, 2000, 'movie');
+
+INSERT INTO "quality_profile" ("name", "cutoff", "upgrade_allowed") 
+VALUES 
+('Standard Definition', 2, TRUE),
+('High Definition', 8, TRUE),
+('Ultra High Definition', 13, FALSE);
+
+INSERT INTO "quality_item" ("quality_id", "name", "allowed", "parent_id")
```
Comment (kasuboski) [inline pkg/storage/sqlite/schema/defaults.sql:25]: Do we need a `quality_item`? `profile_quality_item` could link a profile to the definition directly. I'm probably just unclear what allowed and parent_id are for though. What type of item would `parent_id` point to?

File: pkg/storage/sqlite/sqlite.go [hunk 3]
```diff
@@ -60,29 +63,89 @@ func (s SQLite) CreateIndexer(ctx context.Context, indexer model.Indexers) (int6
 
 // DeleteIndexer deletes a stored indexer given the indexer ID
 func (s SQLite) DeleteIndexer(ctx context.Context, id int64) error {
-	stmt := table.Indexers.DELETE().WHERE(table.Indexers.ID.EQ(sqlite.Int64(id))).RETURNING(table.Indexers.AllColumns)
+	stmt := table.Indexer.DELETE().WHERE(table.Indexer.ID.EQ(sqlite.Int64(id))).RETURNING(table.Indexer.AllColumns)
```
Comment (kasuboski) [inline pkg/storage/sqlite/sqlite.go:66]: Phew I wasn't a plural fan

File: pkg/storage/sqlite/schema/defaults.sql (hunk 2 above)
Comment (kdwils) [reply pkg/storage/sqlite/schema/defaults.sql]: yeah I don't think we do, I ended up removing this as part of the schema and directly referencing the quality definition in the profiles instead

File: pkg/storage/sqlite/sqlite.go (hunk 3 above)
Comment (kdwils) [reply pkg/storage/sqlite/sqlite.go:67]: I swapped your new schema out of plural as well fyi

File: pkg/storage/sqlite/schema/schema.sql (hunk 1 above)
Comment (kdwils) [reply pkg/storage/sqlite/schema/schema.sql]: yes, changed

**Comment by kdwils** [comment]
New example response:

```json
{
  "response": {
    "name": "Standard Definition",
    "qualities": [
      {
        "name": "HDTV-720p",
        "type": "movie",
        "preferredSize": 1999,
        "minSize": 17.1,
        "maxSize": 2000
      },
      {
        "name": "WEBDL-720p",
        "type": "movie",
        "preferredSize": 1999,
        "minSize": 12.5,
        "maxSize": 2000
      }
    ],
    "id": 1,
    "cutoff_quality_id": 2,
    "upgradeAllowed": true
  }
}
```

File: cmd/serve.go [hunk 4]
```diff
@@ -53,12 +54,20 @@ var serveCmd = &cobra.Command{
 			log.Fatal("failed to create prowlarr client", zap.Error(err))
 		}
 
+		defaultSchemas := cfg.Storage.Schemas
+		if _, err := os.Stat(cfg.Storage.FilePath); err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				log.Debug("database does not exist, defaulting table values", zap.Any("schemas", cfg.Storage.TableValueSchemas))
```
Comment (kasuboski) [inline cmd/serve.go:60]: I think we'll probably want more intense version of this at some point (not just no db). But a future problem

File: pkg/manager/manager.go [hunk 5]
```diff
@@ -219,53 +223,52 @@ func (m MediaManager) AddMovieToLibrary(ctx context.Context, request AddMovieReq
...
+
 	return func(r *prowlarr.ReleaseResource) bool {
 		// bytes to megabytes
 		sizeMB := *r.Size >> 20
-		metQuality := MeetsQualitySize(qs, uint64(sizeMB), uint64(*det.Runtime))
-		decision := !metQuality
-		if decision {
-			log.Debugw("rejecting release", "release", r.Title, "metQuality", metQuality, "size", r.Size, "runtime", det.Runtime)
+
+		// items are assumed to be sorted quality so the highest media quality avaiable is selected
```
Comment (kasuboski) [inline pkg/manager/manager.go:252]: avaiable to available

File: pkg/manager/manager.go [hunk 6]
```diff
@@ -219,53 +223,52 @@ func (m MediaManager) AddMovieToLibrary(ctx context.Context, request AddMovieReq
...
+		for _, quality := range profile.Qualities {
+			metQuality := MeetsQualitySize(quality, uint64(sizeMB), uint64(*det.Runtime))
+
+			if metQuality {
+				log.Infow("accepting release", "release", r.Title, "metQuality", metQuality, "size", r.Size, "runtime", det.Runtime)
```
Comment (kasuboski) [inline pkg/manager/manager.go:257]: Was making this info from debug a conscious choice? I had been thinking it's potentially a lot of output

File: pkg/manager/manager.go [hunk 7]
```diff
@@ -321,9 +324,9 @@ func (m MediaManager) AddIndexer(ctx context.Context, request AddIndexerRequest)
 	return indexer, nil
 }
 
-// DeleteIndexerRequest describes what is required to delete an indexer
+// DeleteIndexerRequest request to delete an indexer
 type DeleteIndexerRequest struct {
-	ID *int `json:"id" yaml:"id" mapstructure:"id"`
+	ID *int `json:"id" yaml:"id"`
```
Comment (kasuboski) [inline pkg/manager/manager.go:329]: How come this is a pointer? It's required right

File: pkg/storage/sqlite/schema/schema.sql [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+);
+
+CREATE TABLE IF NOT EXISTS "quality_profile" (
+    "id" INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+    "name" TEXT NOT NULL,
+    "cutoff_quality_id" INTEGER NOT NULL,
+    "upgrade_allowed" BOOLEAN NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS "profile_quality_item" (
```
Comment (kasuboski) [inline pkg/storage/sqlite/schema/schema.sql:26]: This reads kind of odd to me with it being the only one with profile first vs quality.

File: cmd/serve.go (hunk 4 above)
Comment (kdwils) [reply cmd/serve.go:60]: yeah agreed

File: pkg/manager/manager.go (hunk 7 above)
Comment (kdwils) [reply pkg/manager/manager.go:329]: I think I did that so I could check if it was nil vs zero value of 0

File: pkg/storage/sqlite/schema/schema.sql (hunk 8 above)
Comment (kdwils) [reply pkg/storage/sqlite/schema/schema.sql]: yeah I agree, I noticed that myself and ended up not changing it initially. Swapped it to quality_profile_item instead now

File: pkg/manager/manager.go (hunk 6 above)
Comment (kdwils) [reply pkg/manager/manager.go]: nope, reverted. It does spit out a lot of logs if there are a lot of potential releases pulled back from the tracker

File: pkg/manager/manager.go (hunk 5 above)
Comment (kdwils) [reply pkg/manager/manager.go]: fixed

**Review by kasuboski — APPROVED** [review]

**Pull request merged**